### PR TITLE
Unicode

### PR DESCRIPTION
**Pull request by fjebaker**
Following up from #1 (thanks @alberic89) will use this PR to track the changes to get unicode support into fuzzig.

Todo:
- [x] API needs reworking to remove `@panic` (breaking)
- [ ] More rigorous test cases for unicode

### CODE REVIEW

File: src/root.zig [hunk 1]
```diff
@@ -508,9 +539,98 @@ pub const AsciiOptions = struct {
...
+
+        while (iter.next()) |c| {
+            converted_string.append(c.code) catch @panic("Memory error");
+        }
+        return converted_string.toOwnedSlice() catch @panic("Memory error");
+    }
+
+    fn eqlFunc(a: *const UnicodeOptions, h: u21, n: u21) bool {
+        const gcd = GenCatData.init(a.allocator) catch @panic("Memory error");
+        defer gcd.deinit();
```
Comment (fjebaker) [inline src/root.zig]: The `eqlFunc` is called in the inner loop of the alignment solver, so it should ideally have a buffered allocator. As it currently is this will hurt performance with the alloc overhead.

File: src/root.zig [hunk 2]
```diff
@@ -100,6 +107,12 @@ pub fn Algorithm(
 
         impl: Impl,
 
+        const TypeOfCaracter = switch (Impl) {
+            AsciiOptions => u8,
+            UnicodeOptions => u21,
+            else => unreachable,
+        };
```
Comment (fjebaker) [inline src/root.zig]: The `Algorithm` generic struct already has an `ElType` argument, so we can use that instead of having to maintain a switch. In theory I still want downstream users to be able to modify the behaviour of the fuzzy finder without having to modify the source.

File: src/root.zig [hunk 3]
```diff
@@ -463,6 +487,13 @@ pub const AsciiOptions = struct {
 
     penalty_case_mistmatch: i32 = -2,
 
+    /// Don't forget the allocator !!!
+    allocator: Allocator = undefined,
```
Comment (fjebaker) [inline src/root.zig]: Let's pass allocators in from the algorithm struct instead of having the options hold on to them.

File: src/root.zig [hunk 4]
```diff
@@ -508,9 +539,98 @@ pub const AsciiOptions = struct {
...
+        const c = CharacterType.fromUnicode(n, self.allocator);
+
+        return switch (p.roleNextTo(c)) {
+            .Head => scores.bonus_head,
+            .Camel => scores.bonus_camel,
+            .Break => scores.bonus_break,
+            .Tail => scores.bonus_tail,
+        };
+    }
+};
```
Comment (fjebaker) [inline src/root.zig]: These are both essentically identical to the Ascii functions which suggests we should pull them out so we only maintain the implementation once.

File: src/utils.zig [hunk 5]
```diff
@@ -29,6 +32,33 @@ pub const CharacterType = enum {
         };
     }
 
+    pub fn fromUnicode(c: u21, allocator: std.mem.Allocator) CharacterType {
+        const cd = CaseData.init(allocator) catch @panic("Memory error");
+        defer cd.deinit();
+        const gcd = GenCatData.init(allocator) catch @panic("Memory error");
```
Comment (fjebaker) [inline src/utils.zig]: Again, this feels like we should be able to do it with a stack allocated buffer instead of a `std.mem.Allocator` to avoid alloc overheads.

File: src/root.zig [hunk 6]
```diff
@@ -179,34 +195,40 @@ pub fn Algorithm(
                 .score = 0,
             };
 
-            const rows = needle.len;
-            const cols = haystack.len;
+            const haystack_normal = self.impl.convertString(haystack);
+            defer self.allocator.free(haystack_normal);
+
+            const needle_normal = self.impl.convertString(needle);
+            defer self.allocator.free(needle_normal);
```
Comment (fjebaker) [inline src/root.zig]: Maybe use `ArenaAllocator` for the lifetime of a single solve? Then implementations that don't need to allocate don't have to pay the price of a `dupe` to satisfy the free? Or else let the implementation manage the lifetimes?

**Comment by alberic89** [comment]
I will rework on all the change you have suggested.

Meanwhile, I have made other changes to remove the need to know by advance the length of the haystack and needle. You can see them here: https://github.com/alberic89/fuzzig/commit/bb35c1b82e033e22a64a981f7c4bf7dd1fa7bb41

I don't know how I can pull my changes here. Do you want me to make a pull request for each change ?

File: src/root.zig (hunk 3 above)
Comment (alberic89) [reply src/root.zig]: Removed in commit https://github.com/alberic89/fuzzig/commit/3bab90bbb045f787695c80699736ad83483b6235
But many side effects. Maybe they will be attenuated with the futures changes (less need of an allocator).

File: src/root.zig (hunk 2 above)
Comment (alberic89) [reply src/root.zig]: For me, there are two types of solutions:
1. The user should give a standard string (`[]const u8`) to the fuzzy finder, choose the right algo (Ascii or Unicode) depending on the “type” of the text. In this case, `ElType` will always be `u8`, and `TypeOfCaracter` depends on the algo. The algo will make by himself all the conversion.
2. Let the user make the conversions, and the algo will take the type of the strings (`ElType`). The problem is that the user will need a lib to make the conversion, for being able to use fuzzig. And fuzzig will also need a lib (surely the same) to process correctly the data.

The solution for me is the 1, to replace `ElType` by `u8` where it is needed, and use it instead to provide the type that is waited by the `UnicodeOption` (or Ascii).
In this case, `TypeOfCaracter` is no longer needed (so we can delete the switch).

Comment (alberic89) [reply src/root.zig]: I have done it in https://github.com/alberic89/fuzzig/commit/073d10a8f4b2ae945392e1cf059b4af85b3066a1

**Comment by fjebaker** [comment]
@alberic89 I made a couple of API changes. Main thing now is that the Algorithm is owned by the implementation instead of the other way round, which lets the implementation provide the public interface and do things like transforming the input from `u8` to `u21`. It also means the implementations can now access the allocator from the algorithm in their methods too, and provide their own inits and deinits.

Sorry about very likely having introduced a merge conflict for you. If you could rebase your work onto this branch and open PRs into this branch, that would be best.

There's another problem that occured to me; changing from `u8` to `u21` during the input stage means the traceback can be very off. We'd either need to recalculate the offsets into the u8 array, or insist that the unicode input is already `u21`, as I can't think of another way round that.

**Comment by fjebaker** [comment]
I've also had a look at https://github.com/alberic89/fuzzig/commit/bb35c1b82e033e22a64a981f7c4bf7dd1fa7bb41 ; in principle these changes are fine, but the problem is now that the you do the cleanup and the allocation _for each call to `score`_. One of the reasons I had the preallocation was to avoid having to do memory things during score, so you can init a solver and score 100 different fuzzy finds without having to touch memory. What I'd prefer to do here is keep the current interface, but provide a `sscoreResize` or just a `resize` method to let the user make the decision as to whether the buffers are the right size or not, and whether they have the performance budget to check each score's setup or not.

Does that make sense?

**Comment by fjebaker** [comment]
I've added a benchmark target to both the main branch and this one, which you can use to test how it impacts the `Ascii` fuzzy finder with `zig build benchmark`. The benchmark's are naive, but a good coarse estimate.

**Comment by alberic89** [comment]
> I've also had a look at [alberic89@bb35c1b](https://github.com/alberic89/fuzzig/commit/bb35c1b82e033e22a64a981f7c4bf7dd1fa7bb41) ; in principle these changes are fine, but the problem is now that the you do the cleanup and the allocation _for each call to `score`_. One of the reasons I had the preallocation was to avoid having to do memory things during score, so you can init a solver and score 100 different fuzzy finds without having to touch memory. What I'd prefer to do here is keep the current interface, but provide a `sscoreResize` or just a `resize` method to let the user make the decision as to whether the buffers are the right size or not, and whether they have the performance budget to check each score''s setup or not.
> 
> Does that make sense?

Yes, this solution is better. My solution was just “make it working”, but for performance yours is better. I will look on it.

**Comment by fjebaker** [comment]
Sounds good, and thanks for taking a look at this. I'm not married to any particular solution though, so if you find a better solution, please go for it! I am happy to merge breaking changes provided they are worth it :sunglasses:
